Release 2.0.0 to MASTER

lukeindykiewicz opened the pull request.

dilyand commented on .github/workflows/snyk.yml:
```diff
@@ -0,0 +1,20 @@
+name: Snyk
```
Have you tested this @lukeindykiewicz ? I think it won't work without adding the `sbt-dependency-graph` plugin to the project.

To avoid adding one more plugin, you can forego the default Snyk action and instead do it like here: https://github.com/snowplow/scala-maxmind-iplookups/blob/master/.github/workflows/snyk.yml.

lukeindykiewicz replied:
Thanks @dilyand for spotting this!

lukeindykiewicz replied:
fixed in 314c3de

dilyand commented on build.sbt:
```diff
@@ -46,7 +46,7 @@ lazy val commonDependencies = Seq(
   organization  :=  "com.snowplowanalytics",
   name          :=  "snowplow-stream-collector",
-  version       :=  "1.0.1",
+  version       :=  "2.0.0",
```
I think this change should be part of the `Prepare for release` commit. We used to have it as a separate step in the umbrella releases, because it was possible to release a new `R11x` version of `snowplow` and *not* bump the collector. But now that it's in its own repo, we should follow the practice from all our others single-project repos.

benjben replied:
Agreed with Dilyan.

lukeindykiewicz replied:
done

dilyand commented on CHANGELOG:
```diff
@@ -1,3 +1,13 @@
+Release 2.0.0 (2020-08-19)
```
Needs updating.

dilyand commented on CHANGELOG:
```diff
@@ -1,3 +1,13 @@
+Release 2.0.0 (2020-08-19)
+--------------------------
+Bump to 2.0.0 (close #61)
```
Should not be a separate commit.

dilyand reviewed (changes requested):
Hey @lukeindykiewicz , looks good. I've left a couple of comments. Nothing major but I think they need to be addressed before releasing.

I remember reviewing the sqs buffer changes in #48 and the redirect path changes in #65, and I have assumed that this PR is identical to what I have already green-ticked. But let me know if you have made any changes in the interim that need closer attention.

benjben commented on kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala:
```diff
@@ -233,52 +288,150 @@ class KinesisSink private (
+          s"Sending ${failures.size} events from a batch to SQS buffer queue: ${sqs.sqsBufferName}"
+        )
+        putToSqs(sqs, failures)
+        ()
```
Don't we want to wait for the result of `putToSqs` instead of discarding the `Future`? What if all the writes to SQS fail?

lukeindykiewicz replied:
The batch call to SQS is retried up to 10 times. This is already the backup, so unless we want to implement some storage backup for the backup we are pretty much out of options. We don't want to keep it in memory when sqs buffer is enabled, as this is exactly the problem we're fixing with the whole sqs buffer.

benjben replied:
> The batch call to SQS is retried up to 10 times..
If it fails after 10 times, we must have a log message to indicate that we are dropping events.

benjben commented on kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala:
```diff
@@ -233,52 +288,150 @@ class KinesisSink private (
+    }
+
+  private def putToSqs(sqs: SqsClientAndName, batch: List[Event]): Future[Unit] =
+    Future {
```
Why wrapping it in a `Future` ?

lukeindykiewicz replied:
To be fast, we don't have to wait for the result.

lukeindykiewicz commented on .travis/deploy.sh:
```diff
@@ -0,0 +1,6 @@
+
+echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
+
+sbt "project kinesis" docker:publish && sbt "project pubsub" docker:publish && sbt "project kafka" docker:publish && sbt "project nsq" docker:publish && sbt "project stdout" docker:publish
```
Just a convention I think, all repos has it in deploy.sh

lukeindykiewicz commented:
I added two commits that adds the plugin needed by snyk and also yet another bump of jackson was needed, as the bumped version was already vulnerable again.
I've added (to be removed before merging) rc version to check if everything works.

dilyand reviewed (changes requested):
I think this needs three finishing touches:

1.) These two commits (and their corresponding issues) should be merged:
- Add snyk monitor (close #52)
- Add sbt-dependency-graph plugin for snyk (close #69)

The only reason we're adding the plugin is so that snyk monitor will work. It doesn't make sense to have them as separate items.

2.) These two should also be merged (and their corresponding issues):
- Bump vulnerable libs (close #56)
- Bump jackson library due to yet another vulnerability (close #70)

Issues should represent changes since the last published version, but #70 is a change since #56 which is not part of the latest published version.

3.) The CHANGELOG should be updated:
- remove 'Bump to 2.0.0 (close #61)', which is now part of the 'Prepare for release'  commit
- reflect the final state of commits after the merges in 1 and 2
- a milestone should be created with the issues we are closing in this release
- redundant issues (like 70, 69, 61) should be removed.

To make it easier to track progress:

- [x] merge snyk commits
- [x] merge lib bump commits
- [ ] milestone 2.0.0
- [ ] CHANGELOG cleanup
- [ ] final check of 1:1:1 sync between issue tracker, CHANGELOG and commit history

benjben reviewed (changes requested):
We are missing a log message to indicate that we are dropping events after 10 unsuccessful attempt to write to SQS

lukeindykiewicz commented:
@benjben The log messages shows what failed and what succeeded. Nothing changes since this PR was merged: https://github.com/snowplow/stream-collector/pull/48
Around this place: https://github.com/snowplow/stream-collector/pull/66/files#diff-67fdfd8fe30fa650f6a58346af2157feR402

benjben commented:
@lukeindykiewicz still we don't know when the collector is dropping messages.

dilyand commented:
> @lukeindykiewicz still we don't know when the collector is dropping messages.

Ben is right, we can potentially see 9 log messages for SQS failures but the write can succeed on the 10th attempt. In order to know for sure if we're dropping events, we'll need to count the failures for each batch. It would be hard to build reliable monitoring and alerting on top of that.

dilyand commented on kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala:
```diff
@@ -138,10 +175,15 @@ class KinesisSink private (
+  // but we encode the message with Base64 for SQS, so the limit drops to 192k
+  val SqsLimit = 192000 // 256000 / 4 * 3
+  val KinesisLimit = 1000000
+  override val MaxBytes = if (maybeSqs.isDefined) SqsLimit else KinesisLimit
```
There is a big difference between the MaxBytes allowed by Kinesis and SQS, which means there's a tradeoff on using surge protection -- your events must be way leaner. Not something to address in this PR but we should make it very clear in all documentation and comms (with customers / users but also Support / Tech Ops / Product).

lukeindykiewicz replied:
That is reflected in announcement and blog post. Thanks for good point.

dilyand reviewed (approved):
LGTM!

If possible, we should test this with a new rc to ensure nothing broke but also to see if we can make it drop events and log that.

Also, just as a point of process, the issues in the 2.0.0 milestone must stay open. They are auto-closed when the release is merged to `master` (that is the point of the `close #XX` in the commit messages). This is documented here: https://github.com/snowplow-incubator/engineering-resources/blob/master/release-process.md.

lukeindykiewicz commented:
> If possible, we should test this with a new rc to ensure nothing broke but also to see if we can make it drop events and log that.

definitely

> Also, just as a point of process, the issues in the 2.0.0 milestone must stay open. They are auto-closed when the release is merged to master (that is the point of the close #XX in the commit messages). This is documented here: https://github.com/snowplow-incubator/engineering-resources/blob/master/release-process.md.

The issues were closed when the feature PR was merged into develop from feature branch I believe. The gitflow workflow is not 100% compatible with what is described in the release-process doc.

dilyand commented:
> The issues were closed when the feature PR was merged into develop from feature branch I believe. 

Should we change the default branch then? Up until a release takes place, there is a possibility to discover bugs that require the issue to be reopened.

benjben commented:
We should definitely change the default branch to `master` ! @chuwy do you have the power to do it ?

lukeindykiewicz commented:
I have the admin rights here, so I can change it.

benjben commented on kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala:
```diff
@@ -233,52 +288,160 @@ class KinesisSink private (
+      sqsBatchEntries.grouped(MaxSqsBatchSize).foreach { batchEntryGroup =>
+        sendToSqsWithRetry(sqs, batchEntryGroup, MaxRetries).transform {
+          case failure @ Failure(ex) =>
+            log.info(s"Sending to sqs failed with exception: $ex")
```
In this case, are events also dropped? If yes it would be great to include it in the log line.

benjben reviewed (approved):
:+1:
